src/client/Client.cc

```diff
@@ -11707,6 +11707,36 @@ int64_t Client::_write(Fh *f, int64_t offset, uint64_t size, const char *buf,
   return r;
 }
 
+int64_t Client::_maybe_finish_context(int64_t r, Context *onfinish) {
+  /* There are two scenarios with each having two cases to handle here
+  1) async io
+    1.a) r == 0:
+      async call in progress, the context will be automatically invoked,
+      so just return the retval (i.e. zero).
+    1.b) r < 0:
+      There was an error; no context completion should've took place so
+      complete the context with retval followed by returning zero to the
+      caller.
+  2) sync io
+    2.a) r >= 0:
+      sync call success; return the no. of bytes read/written.
+    2.b) r < 0:
+      sync call failed; return the errno. */
+
+  ceph_assert(ceph_mutex_is_locked_by_me(client_lock));
```

Comment:
please stop inserting ceph_assert(...) calls into fast paths--or really, any paths.  this is really a problem for performance over increasingly larger areas of the ceph codebase.  ceph_assert(...) is not compiled out as an ordinary assert(...) is, and should not be used to assert facts that should be manifest in the code.  if you're trying to inform the reader that there is a hidden invariant, write a comment (/* ASSERT: we hold foo mutex */);  if use of this helper function doesn't make that manifest, then maybe there is a problem with the decomposition.

Reply:
> if you're trying to inform the reader that there is a hidden invariant, write a comment (/* ASSERT: we hold foo mutex */); if use of this helper function doesn't make that manifest, then maybe there is a problem with the decomposition.

the intention is to make sure the client lock is held so commenting something like that might not help.